src/transformers/pipelines/base.py

```diff
@@ -739,65 +824,69 @@ def check_model_type(self, supported_models: Union[List[str], dict]):
-                    model_input = self._parse_and_tokenize(input_, padding=False, *args, **kwargs)
-                    value = self._forward(model_input)
-                    values.append(value.squeeze(0))
-            else:
-                model_input = self._parse_and_tokenize(inputs, padding=False, *args, **kwargs)
-                values = self._forward(model_input)
-            return values
+    @abstractmethod
+    def set_parameters(self, **pipeline_parameters):
+        raise NotImplementedError("set_parameters not implemented")
 
-    def _forward(self, inputs, return_tensors=False):
-        """
-        Internal framework specific forward dispatching
+    @abstractmethod
+    def preprocess(self, input_: Any, **preprocess_parameters) -> Dict[str, GenericTensor]:
+        raise NotImplementedError("preprocess not implemented")
 
-        Args:
-            inputs: dict holding all the keyword arguments for required by the model forward method.
-            return_tensors: Whether to return native framework (pt/tf) tensors rather than numpy array
+    @abstractmethod
+    def postprocess(self, model_outputs: ModelOutput, **postprocess_parameters) -> Any:
+        raise NotImplementedError("postprocess not implemented")
 
-        Returns:
-            Numpy array
-        """
-        # Encode for forward
+    @abstractmethod
+    def forward(self, input_tensors: Dict[str, GenericTensor], **forward_parameters) -> ModelOutput:
+        raise NotImplementedError("postprocess not implemented")
+
+    def infer_forward(self, model_inputs, *args, **kwargs):
         with self.device_placement():
             if self.framework == "tf":
-                # TODO trace model
-                predictions = self.model(inputs.data, training=False)[0]
-            else:
+                model_inputs["training"] = False
+                model_outputs = self.forward(model_inputs)
+            elif self.framework == "pt":
                 with torch.no_grad():
-                    inputs = self.ensure_tensor_on_device(**inputs)
-                    predictions = self.model(**inputs)[0].cpu()
-
-        if return_tensors:
-            return predictions
+                    model_inputs = self.ensure_tensor_on_device(**model_inputs)
+                    model_outputs = self.forward(model_inputs)
+                    model_inputs = self.ensure_tensor_on_cpu(model_outputs)
+            else:
+                raise ValueError(f"Framework {self.framework} is not supported")
+        return model_outputs
+
+    def get_iterator(self, inputs, num_workers: int):
+        os.environ["TOKENIZERS_PARALLELISM"] = "false"
+        dataset = PipelineDataset(inputs, self.preprocess)
+        dataloader = DataLoader(dataset, num_workers=num_workers, batch_size=1, collate_fn=collate_fn)
+        model_iterator = PipelineIterator(dataloader, self.infer_forward)
+        final_iterator = PipelineIterator(model_iterator, self.postprocess)
+        return final_iterator
```

Comment:
This could also return `list(final_iterator)` to return the computed values

Reply:
That would defeat the principle of having an iterator (as calling `list` for force to generate all outputs first leading to RAM issues on very large datasets at the very least).
It would also be great to have infinite generators be supported (like waiting on a queue that would be great in an API :))

Reply:
The idea is to return something you can iterate on, on an item basis.

`for item in iterator`

Reply:
Ah, my bad :)